hist/histv7/inc/ROOT/RHistImpl.hxx

```diff
@@ -227,122 +257,467 @@ namespace Internal {
+struct RComputeLocalBinsRaw;
+
+template <int NDIMS, typename BINS, class AXES, class BINTYPE>
+struct RComputeLocalBinsRaw<-1, NDIMS, BINS, AXES, BINTYPE> {
+   void operator()(BINS & /*virtualBins*/, const AXES & /*axes*/, int /*zeroBasedGlobalBin*/, BINTYPE /*GetNBinType*/) const
+   {}
+};
 
-// Break recursion.
-template <class AXES>
-struct RFillIterRange<-1, AXES> {
-   void operator()(Hist::AxisIterRange_t<std::tuple_size<AXES>::value> & /*range*/, const AXES & /*axes*/,
-                   const std::array<Hist::EOverflow, std::tuple_size<AXES>::value> & /*over*/) const
+template <int I, int NDIMS, typename BINS, class AXES, class BINTYPE>
+struct RComputeLocalBinsRaw {
+   void operator()(BINS &virtualBins, const AXES &axes, int zeroBasedGlobalBin, BINTYPE GetNBinType) const
+   {
+      constexpr const int thisAxis = NDIMS - I - 1;
+      virtualBins[thisAxis] = zeroBasedGlobalBin % (std::get<thisAxis>(axes).*GetNBinType)();
+      RComputeLocalBinsRaw<I - 1, NDIMS, BINS, AXES, BINTYPE>()(virtualBins, axes, zeroBasedGlobalBin / (std::get<thisAxis>(axes).*GetNBinType)(), GetNBinType);
+   }
+};
+
+/// Recursively computes a zero-based global bin index, given...
+///
+/// - A set of zero-based per-axis bin indices
+/// - The number of considered bins on each axis (can be either `GetNBinsNoOver`
+///   or `GetNBins` depending on what you are trying to do)
+/// - A policy of treating all bins qs regular (i.e. no negative indices)
+template <int I, int NDIMS, typename BINS, class AXES, class BINTYPE>
+struct RComputeGlobalBinRaw;
+
+template <int NDIMS, typename BINS, class AXES, class BINTYPE>
+struct RComputeGlobalBinRaw<-1, NDIMS, BINS, AXES, BINTYPE> {
+   int operator()(int globalVirtualBin, const AXES & /*axes*/, const BINS & /*zeroBasedLocalBins*/, int /*binSize*/, BINTYPE /*GetNBinType*/) const
+   {
+      return globalVirtualBin;
+   }
+};
+
+template <int I, int NDIMS, typename BINS, class AXES, class BINTYPE>
+struct RComputeGlobalBinRaw {
+   int operator()(int globalVirtualBin, const AXES &axes, const BINS &zeroBasedLocalBins, int binSize, BINTYPE GetNBinType) const
+   {
+      constexpr const int thisAxis = NDIMS - I - 1;
+      globalVirtualBin += zeroBasedLocalBins[thisAxis] * binSize;
+      binSize *= (std::get<thisAxis>(axes).*GetNBinType)();
+      return Internal::RComputeGlobalBinRaw<I - 1, NDIMS, BINS, AXES, BINTYPE>()(globalVirtualBin, axes, zeroBasedLocalBins, binSize, GetNBinType);
+   }
+};
+
+/// Recursively converts zero-based virtual bins where the underflow bin
+/// has index `0` and the overflow bin has index `N+1` where `N` is the axis'
+/// number of regular bins, to the standard `kUnderflowBin`/`kOverflowBin` for under/overflow
+/// bin indexing convention.
+///
+/// For growable axes, add 1 from regular indices so that the indexing
+/// convention remains zero-based (this means that there will be no "holes" in
+/// global binning, which matters more than the choice of regular index base).
```

Comment:
```suggestion
/// Recursively converts zero-based virtual bins where the underflow bin (if any)
/// has index `0` and the overflow bin has index `N+1` where `N` is the axis'
/// number of regular bins, to the standard `kUnderflowBin`/`kOverflowBin` for under/overflow
/// bin indexing convention.
///
/// For growable axes, must add 1 to go back to standard indices as their virtual
/// indexing convention is also zero-based, with zero designating the first regular bin.
```

(The comment about leaving no "holes" is probably best left on `RLocalBinsToVirtualBins` only.)